### DISCUSSION TITLE
Why does l2arc take space in RAM rather than on the l2arc device itself?

### DISCUSSION THREAD

**Discussion by the-omega-zeta** [openzfs/zfs#12479] [Q&A]
Why does l2arc take space in RAM rather than on the l2arc device itself? Does it lose cache effectiveness by doing so? Surely given the cache heirarchy it would still be better than spinning disk, right?

**Comment by IvanVolosyuk** [comment]
ZFS has to know if the object exists in L2ARC to read it from there. It
will be too expensive to read the L2ARC device just to understand if the
object is there. Thus, L2ARC uses some ARC RAM to store the dictionary of
what's available in the l2ARC device and where exactly. Having too large
L2ARC devices may cause large portion of ARC to be used by the dictionary
and be unavailable for ARC (first layer cache). This may actually slow the
system down. Different workloads may need different sizes of L2ARC. When
unsure don't add L2ARC larger than 5x of the system ram.

**Comment by grahamperrin** [comment]
From [OpenZFS: All about the cache vdev or L2ARC | Klara Inc.](https://klarasystems.com/articles/openzfs-all-about-l2arc/) (undated): 

> … exactly how much RAM is necessary to index the L2ARC, …
